### PR TITLE
feat(scheduler): admission data layer — inflight, queue, slots (PR 2 M2a)

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

PR 2 (priority scheduler) is split into milestones. M1 (#1541) landed the vocabulary types (`Class`, configs, `TenantPolicyResolver`). This PR is **M2a**, the first half of M2 (scheduler core): the data-layer types the engine will own — in-flight bookkeeping, per-class waiter queue, and packed-CAS slot accounting.

M2b will follow with the engine itself (`PriorityScheduler::new`, `admit`, dispatcher, capacity-watch integration).

### Solution

Three small, self-contained modules with full unit-test coverage. **No production wiring** — these types are reachable from `crate::middleware::scheduler::*` but no call site constructs them yet. Behavior on `main` is unchanged.

## Changes

Three commits, one per task:

1. **`feat(scheduler): InflightHandle with TTFT/preempt CAS sentinels`** — `InflightHandle` tracks a single admitted request via one `AtomicU64` (`first_byte_at`). Three states encoded in one word: `0` = no TTFT yet, `u64::MAX` = preempted (sentinel), anything else = millis-since-admit at TTFT. Two symmetric CAS methods (`try_mark_preempted`, `try_mark_first_byte`) resolve the race between the scheduler firing a preempt and the response body emitting its first frame — whichever wins the CAS dictates the outcome, the loser respects the winner. `try_mark_first_byte` clamps to `u64::MAX - 1` so a TTFT measurement can never collide with the sentinel.

2. **`feat(scheduler): FIFO per-class queue + ClassQueue trait`** — `Waiter` (class + queued-at instant + cancel token) and `ClassQueue` trait (`try_enqueue` / `pop_eligible` / `head_age` / `depth` / `drop_cancelled_head`). v1 impl is `FifoClassQueue { Mutex<VecDeque<Waiter>>, max }`. Trait shape lets a future per-tenant fair-queue swap in without touching the dispatcher. `drop_cancelled_head` GCs walkaway clients so they don't consume a slot when their turn comes.

3. **`feat(scheduler): packed-CAS slot accounting with reservation guard`** — `SlotPool` packs four `u16` per-class in-flight counts into one `AtomicU64`. A single compare-and-swap both verifies availability under the reservation guard and claims the slot, so concurrent admissions to different classes can't race and double-spend a higher-class reservation. `try_acquire` honors reservations; `try_acquire_ignoring_reservations` is the starvation-override path. Released slots saturate at zero (defensive against stray double-releases).

## Test Plan

Each task: failing test → verify FAIL → implement → verify PASS → commit. 26 new unit tests:

- **inflight (7)**: preempt CAS succeeds-then-fails, TTFT CAS succeeds-then-fails, mutual exclusion in both directions (TTFT loses if preempt already won, and vice versa), sentinel-collision clamp at `u64::MAX - 1`, initial state is zero.
- **queue (8)**: enqueue fills to capacity then rejects with `Err(waiter)`, pop FIFO order, head_age tracks wall-clock wait, depth reflects mutations, `drop_cancelled_head` removes cancelled head only and is a no-op on live head / empty queue.
- **slots (11)**: pack/unpack round-trip across 6 sample vectors including `[u16::MAX; 4]`, `slots_available_to` honors higher-class reservations / ignores same-class / collapses 1:1 on consumed reservation, `try_acquire` respects guard / returns false when full without spinning, `try_acquire_ignoring_reservations` bypasses guard, release saturates at zero, and a **concurrent contention test**: 200 threads racing against capacity=64 admit exactly 64 (no over- or under-admission).

```bash
$ cargo test --package smg --lib middleware::scheduler
... 55 passed; 0 failed
```

```bash
$ cargo clippy --package smg --all-targets --all-features -- -D warnings
... Finished `dev` profile, no warnings
```

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes (enforced by pre-commit hook)
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [x] (Optional) Documentation updated — see `.claude/docs/scheduler/02-priority-scheduler-plan.md` (in #1541)
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

## Summary by CodeRabbit

* **New Features**
  * Enhanced scheduler: precise in-flight tracking for admitted requests and a preemption marker.
  * Per-class FIFO admission queues with cancellation handling, wait-age reporting, and bounded capacity.
  * Slot-based capacity manager with per-class reservations, reservation-aware admissions, and robust release behavior.
* **Tests**
  * Comprehensive unit tests covering booking, queue behavior, slot semantics, edge cases, and contention.

[Review Change Stack](https://app.coderabbit.ai/change-stack/lightseekorg/smg/pull/1546?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: model_gateway/src/middleware/scheduler/inflight.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    /// (or any call after preemption was marked) return `false`.
+    ///
+    /// `now_ms` is clamped to `u64::MAX - 1` so a TTFT measurement can
+    /// never collide with the preempt sentinel.
+    pub fn try_mark_first_byte(&self, now_ms: u64) -> bool {
+        let value = now_ms.min(Self::PREEMPTED_SENTINEL - 1);
+        self.first_byte_at
+            .compare_exchange(0, value, Ordering::AcqRel, Ordering::Acquire)
+            .is_ok()
+    }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/middleware/scheduler/inflight.rs:71]: Fixed in `61a4d4f4`. Clamped to `[1, PREEMPTED_SENTINEL - 1]`. Added regression test `test_ttft_at_zero_ms_does_not_collide_with_unset_sentinel` that calls `try_mark_first_byte(0)` and asserts a subsequent `try_mark_preempted` is locked out.

File: model_gateway/src/middleware/scheduler/queue.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    fn depth(&self) -> usize {
+        self.waiters.lock().len()
+    }
+
+    fn drop_cancelled_head(&self) {
+        let mut guard = self.waiters.lock();
+        if guard.front().is_some_and(|w| w.cancel.is_cancelled()) {
+            guard.pop_front();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/middleware/scheduler/queue.rs:110]: Fixed in `58bf5fb0`. Switched to a `while` loop; one drain pass clears the whole run in one lock acquisition. Updated the trait doc to match. New test `test_drop_cancelled_head_drains_run_in_single_call` covers the multi-head case.

File: model_gateway/src/middleware/scheduler/slots.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+                cur,
+                pack(counts),
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            ) {
+                Ok(_) => return,
+                Err(observed) => cur = observed,
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/middleware/scheduler/slots.rs:179]: Fixed in `85a1a403`. `release` now early-returns when the lane reads zero, avoiding the unchanged-value CAS write and its cache-coherence traffic. Existing `test_release_saturates_at_zero` still validates the functional behavior.

File: model_gateway/src/middleware/scheduler/inflight.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    /// Attempt to mark the first response byte at `now_ms` millis-since-
+    /// admit. Returns `true` on the first successful CAS; subsequent calls
+    /// (or any call after preemption was marked) return `false`.
+    ///
+    /// `now_ms` is clamped to `u64::MAX - 1` so a TTFT measurement can
+    /// never collide with the preempt sentinel.
+    pub fn try_mark_first_byte(&self, now_ms: u64) -> bool {
+        let value = now_ms.min(Self::PREEMPTED_SENTINEL - 1);
+        self.first_byte_at
+            .compare_exchange(0, value, Ordering::AcqRel, Ordering::Acquire)
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/middleware/scheduler/inflight.rs]: Fixed in `61a4d4f4`. Clamped TTFT to `[1, PREEMPTED_SENTINEL - 1]` so `now_ms == 0` stores `1` rather than the unset sentinel. Regression test `test_ttft_at_zero_ms_does_not_collide_with_unset_sentinel` asserts mutual exclusion holds after a zero-ms TTFT.

File: model_gateway/src/middleware/scheduler/inflight.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            .is_ok()
+    }
+
+    /// Attempt to mark the first response byte at `now_ms` millis-since-
+    /// admit. Returns `true` on the first successful CAS; subsequent calls
+    /// (or any call after preemption was marked) return `false`.
+    ///
+    /// `now_ms` is clamped to `u64::MAX - 1` so a TTFT measurement can
+    /// never collide with the preempt sentinel.
+    pub fn try_mark_first_byte(&self, now_ms: u64) -> bool {
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/middleware/scheduler/inflight.rs:66]: Fixed in `61a4d4f4`. Applied the suggested `now_ms.max(1).min(PREEMPTED_SENTINEL - 1)` (expressed as `clamp(1, …)`). Regression test covers `try_mark_first_byte(0)` followed by `try_mark_preempted()` and asserts the latter returns `false`.

File: model_gateway/src/middleware/scheduler/inflight.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        let value = now_ms.min(Self::PREEMPTED_SENTINEL - 1);
+        self.first_byte_at
+            .compare_exchange(0, value, Ordering::AcqRel, Ordering::Acquire)
+            .is_ok()
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/middleware/scheduler/inflight.rs:70]: Fixed in `61a4d4f4`. Verified the CAS semantics — `compare_exchange(0, 0, …)` does succeed (yes, that was the bug). Clamped TTFT to `[1, PREEMPTED_SENTINEL - 1]` so the stored value can never collide with the unset sentinel. New test `test_ttft_at_zero_ms_does_not_collide_with_unset_sentinel` is the regression guard.

File: model_gateway/src/middleware/scheduler/inflight.rs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+/// State encoding for `first_byte_at`:
+///
+/// | Value         | Meaning                                                   |
+/// |---------------|-----------------------------------------------------------|
+/// | `0`           | No first-byte emitted yet and not preempted (initial)     |
+/// | `u64::MAX`    | Scheduler preempted this request before TTFT (sentinel)   |
+/// | anything else | Millis-since-admit of the first response byte (TTFT)      |
+///
+/// `try_mark_first_byte` clamps `now_ms` to `u64::MAX - 1` so a TTFT
+/// measurement can never collide with the preempt sentinel.
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/middleware/scheduler/inflight.rs]: Good catch — fixed in `a3c571f3`. Struct doc now says `[1, u64::MAX - 1]` so both bounds are described.
